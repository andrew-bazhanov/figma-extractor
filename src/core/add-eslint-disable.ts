const maxLineRuleComment = '/* eslint-disable max-lines */';

const namingConventionRuleComment = '/* eslint-disable @typescript-eslint/naming-convention */';

export const addEslintDisableAtTheTopOfText = (
  content: string,
  options: {
    disabledMaxLines: boolean;
    disabledTypescriptNamingConvention: boolean;
  },
) => {
  return `
    ${options.disabledMaxLines ? maxLineRuleComment : ''}
    ${options.disabledTypescriptNamingConvention ? namingConventionRuleComment : ''}
    ${content}
    `;
};
