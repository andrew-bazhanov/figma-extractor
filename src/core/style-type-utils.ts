import { StyleType } from 'figma-js';

// eslint-disable-next-line @typescript-eslint/naming-convention
type StyleTypePredicate = (item: { style_type: StyleType }) => boolean;

const getStyleTypePredicate = (styleType: StyleType): StyleTypePredicate => {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  return ({ style_type }) => style_type === styleType;
};

export const styleTypeUtils = {
  isFill: getStyleTypePredicate('FILL'),
  isEffect: getStyleTypePredicate('EFFECT'),
  isText: getStyleTypePredicate('TEXT'),
};
