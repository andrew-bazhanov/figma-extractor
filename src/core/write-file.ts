import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

const defaultOptions = {
  useEslintDisabledRules: true,
};

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeFile = (
  template: string,
  filePath: string,
  options?: {
    useEslintDisabledRules?: boolean;
  },
) => {
  const dir = path.dirname(filePath);
  const useEslintDisabledRules = { ...defaultOptions, ...options }.useEslintDisabledRules;

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  const content = useEslintDisabledRules
    ? `
    ${fileLengthRuleDisableComment}
    ${namingConventionRuleDisableComment}
    ${template}
    `
    : template;

  try {
    fs.writeFileSync(filePath, content);
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
