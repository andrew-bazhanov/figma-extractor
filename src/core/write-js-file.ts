import fs from 'fs';
import path from 'path';

const fileLengthRuleDisableComment = '/* eslint-disable max-lines */';

const namingConventionRuleDisableComment =
  '/* eslint-disable @typescript-eslint/naming-convention */';

// NOTE: temporarily changed writeFile to writeFileSync to make testing easier:
// it is unclear how to test callback version without introducing promises.
// But even with promises, the logic seems to "just write" contents without returning the promise itself,
// which is confusing a bit, because it is impossible to handle promise fulfillment.
export const writeJsFile = (template: string, filePath: string) => {
  const dir = path.dirname(filePath);

  // Create directory in advance if there is no one
  if (dir && !fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }

  try {
    fs.writeFileSync(
      filePath,
      `
      ${fileLengthRuleDisableComment}
      ${namingConventionRuleDisableComment}
      ${template}
      `,
    );
    console.log(`Generate ${filePath} is completed`);
  } catch {
    console.log(`Formatting ${filePath} file...`);
  }
};
