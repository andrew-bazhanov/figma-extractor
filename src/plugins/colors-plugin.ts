import path from 'path';

import { getColorName } from '../lib/color/get-color-name/get-color-name';
import { getColorStyles } from '../lib/get-color-styles';
import { stringifyRecordsWithSort } from '../lib/stringify';
import { Plugin } from './types';

export const colorsPlugin: Plugin = (
  { config, styleTypeUtils, writeFile, addEslintDisableRules },
  { styleMetadata, fileNodes },
) => {
  const metaColors = styleMetadata.filter(styleTypeUtils.isFill);

  if (!config?.styles?.colors?.disabled) {
    const colors = getColorStyles(
      metaColors,
      fileNodes,
      config?.styles?.colors?.keyName ?? getColorName,
    );

    const colorTemplate = `module.exports = ${stringifyRecordsWithSort(colors)};`;

    writeFile(
      addEslintDisableRules(colorTemplate, ['disable-max-lines']),
      path.join(config?.styles?.exportPath || '', 'colors.js'),
    );
  }
};

colorsPlugin.pluginName = 'colors';
