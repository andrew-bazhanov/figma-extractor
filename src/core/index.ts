import { ClientInterface, FileNodesResponse, FullStyleMetadata } from 'figma-js';

import { Plugin } from '../plugins';
import { runFormattingFile } from './run-formatting-file';
import { styleTypeUtils } from './style-type-utils';
import { writeJsFile } from './write-js-file';

export interface FigmaData {
  figmaClient: ClientInterface;
  styleMetadata: readonly FullStyleMetadata[];
  fileNodes: FileNodesResponse;
}

export interface Core {
  config: Config;
  rootPath: string;
  plugins: Plugin[];
  styleTypeUtils: typeof styleTypeUtils;
  writeJsFile: typeof writeJsFile;
  runFormattingFile: typeof runFormattingFile;
}

export const createCore = (args: { config: Config; rootPath: string; plugins: Plugin[] }): Core => {
  return {
    ...args,
    styleTypeUtils,
    writeJsFile,
    runFormattingFile: runFormattingFile,
  };
};
