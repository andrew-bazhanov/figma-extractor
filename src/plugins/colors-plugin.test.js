import { createCore } from '../core';
import { fileNodes } from '../utils/generate-styles/fixtures/file-nodes';
import { styleMetadata } from '../utils/generate-styles/fixtures/style-metadata';
import { colorsPlugin } from './colors-plugin';

describe('colorsPlugin', () => {
  it('should write js file and run formatting tool', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          colors: {},
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeJsFile = jest.fn();
    core.runFormattingFile = jest.fn();

    colorsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes });

    expect(core.writeJsFile).toHaveBeenCalled();
    expect(core.writeJsFile.mock.calls[0][1]).toBe('/export-path/colors.js');

    expect(core.runFormattingFile).toHaveBeenCalled();
    expect(core.runFormattingFile).toHaveBeenCalledWith('/export-path/colors.js');
  });

  it('should write generated date for default config', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          colors: {},
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeJsFile = jest.fn();
    core.runFormattingFile = jest.fn();

    colorsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes });

    expect(core.writeJsFile).toHaveBeenCalled();
    expect(core.writeJsFile.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should be skipped due to disabled field', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          colors: {
            disabled: true,
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeJsFile = jest.fn();
    core.runFormattingFile = jest.fn();

    colorsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes });

    expect(core.writeJsFile).not.toHaveBeenCalled();
    expect(core.runFormattingFile).not.toHaveBeenCalled();
  });

  it('should create data with custom keyName function', () => {
    const core = createCore({
      config: {
        styles: {
          exportPath: '/export-path/',
          colors: {
            keyName: name => name + '__extra',
          },
        },
      },
      plugins: [],
      rootPath: '/root-path',
    });

    core.writeJsFile = jest.fn();
    core.runFormattingFile = jest.fn();

    colorsPlugin(core, { styleMetadata: styleMetadata.styles, fileNodes });

    expect(core.writeJsFile).toHaveBeenCalled();
    expect(core.writeJsFile.mock.calls[0][0]).toMatchSnapshot();
  });
});
