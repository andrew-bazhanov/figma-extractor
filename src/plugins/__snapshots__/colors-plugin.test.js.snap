// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorsPlugin should create data with custom keyName function 1`] = `"module.exports = {"colorful/text/txt900__extra":"#00ffff","dark/text/txt700__extra":"#00ff00","dark/text/txt900__extra":"#000000","light/text/txt900__extra":"#ffffff","monochrome/text/txt900__extra":"#ff00ff","text/txt800__extra":"#3fc3fc3fc","text/txt900__extra":"#2a3fff"};"`;

exports[`colorsPlugin should write generated date for default config 1`] = `"module.exports = {"colorful/text/txt900":"#00ffff","dark/text/txt700":"#00ff00","dark/text/txt900":"#000000","light/text/txt900":"#ffffff","monochrome/text/txt900":"#ff00ff","text/txt800":"#3fc3fc3fc","text/txt900":"#2a3fff"};"`;
